File: news_server/modules/source/infrastructure/getters/baidu.ts
/**
 * 百度热搜新闻源获取器
 * 从 newsnow/server/sources/baidu.ts 迁移
 */
import type { SourceGetter } from '../../domain/getter/getter'
import { NewsItem } from '../../domain/newsitem/entity'

interface BaiduResponse {
  data: {
    cards: {
      content: {
        isTop?: boolean
        word: string
        rawUrl: string
        desc?: string
      }[]
    }[]
  }
}

export const baiduGetter: SourceGetter = async () => {
  const response = await fetch('https://top.baidu.com/board?tab=realtime')
  if (!response.ok) {
    throw new Error(`Failed to fetch Baidu: ${response.statusText}`)
  }

  const rawData = await response.text()
  const jsonMatch = rawData.match(/<!--s-data:(.*?)-->/s)
  
  if (!jsonMatch || !jsonMatch[1]) {
    throw new Error('Failed to parse Baidu data')
  }

  const data: BaiduResponse = JSON.parse(jsonMatch[1])

  return data.data.cards[0].content
    .filter((k) => !k.isTop)
    .map((k) =>
      NewsItem.fromEntity({
        id: k.rawUrl,
        title: k.word,
        url: k.rawUrl,
        extra: {
          hover: k.desc,
        },
      })
    )
}

File: modules/source/infrastructure/getters/baidu.go
package getters

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
)

func (r *Registry) baidu(ctx context.Context) ([]Item, error) {
	body, err := r.get(ctx, "https://top.baidu.com/board?tab=realtime", nil)
	if err != nil {
		return nil, err
	}
	re := regexp.MustCompile(`(?s)<!--s-data:(.*?)-->`)
	m := re.FindSubmatch(body)
	if len(m) < 2 {
		return nil, fmt.Errorf("baidu payload not found")
	}
	var res struct {
		Data struct {
			Cards []struct {
				Content []struct {
					IsTop  bool   `json:"isTop"`
					Word   string `json:"word"`
					RawURL string `json:"rawUrl"`
					Desc   string `json:"desc"`
				} `json:"content"`
			} `json:"cards"`
		} `json:"data"`
	}
	if err := json.Unmarshal(m[1], &res); err != nil {
		return nil, err
	}
	var out []Item
	if len(res.Data.Cards) == 0 {
		return out, nil
	}
	for _, k := range res.Data.Cards[0].Content {
		if k.IsTop {
			continue
		}
		out = append(out, Item{ID: k.RawURL, Title: k.Word, URL: k.RawURL, Extra: map[string]any{"hover": k.Desc}})
	}
	return out, nil
}
